src/main/java/com/commercetools/sync/services/impl/TaxCategoryServiceImpl.java

```diff
@@ -1,60 +1,78 @@
 import java.util.concurrent.CompletionStage;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Consumer;
 
-import static java.lang.String.format;
-import static org.apache.commons.lang3.StringUtils.isBlank;
+import static java.util.Collections.singleton;
 
-public final class TaxCategoryServiceImpl implements TaxCategoryService {
-    private final ProductSyncOptions syncOptions;
-    private final Map<String, String> keyToIdCache = new ConcurrentHashMap<>();
+public final class TaxCategoryServiceImpl extends BaseService<TaxCategoryDraft, TaxCategory, BaseSyncOptions,
+    TaxCategoryQuery, TaxCategoryQueryModel, TaxCategoryExpansionModel<TaxCategory>>
+    implements TaxCategoryService {
 
-    public TaxCategoryServiceImpl(@Nonnull final ProductSyncOptions syncOptions) {
-        this.syncOptions = syncOptions;
+    public TaxCategoryServiceImpl(@Nonnull final TaxCategorySyncOptions syncOptions) {
+        super(syncOptions);
     }
 
     @Nonnull
     @Override
     public CompletionStage<Optional<String>> fetchCachedTaxCategoryId(@Nullable final String key) {
-        if (isBlank(key)) {
-            return CompletableFuture.completedFuture(Optional.empty());
-        }
-        if (keyToIdCache.isEmpty()) {
-            return fetchAndCache(key);
-        }
-        return CompletableFuture.completedFuture(Optional.ofNullable(keyToIdCache.get(key)));
+        return fetchCachedResourceId(key);
     }
 
     @Nonnull
-    private CompletionStage<Optional<String>> fetchAndCache(@Nonnull final String key) {
-        final Consumer<List<TaxCategory>> taxCategoryPageConsumer = taxCategoryPage ->
-            taxCategoryPage.forEach(taxCategory -> {
-                final String fetchedTaxCategoryKey = taxCategory.getKey();
-                final String id = taxCategory.getId();
-                if (StringUtils.isNotBlank(fetchedTaxCategoryKey)) {
-                    keyToIdCache.put(fetchedTaxCategoryKey, id);
-                } else {
-                    syncOptions.applyWarningCallback(format("TaxCategory with id: '%s' has no key set. Keys are"
-                        + " required for taxCategory matching.", id));
-                }
-            });
-
-        return CtpQueryUtils.queryAll(syncOptions.getCtpClient(), TaxCategoryQuery.of(), taxCategoryPageConsumer)
-                            .thenApply(result -> Optional.ofNullable(keyToIdCache.get(key)));
+    @Override
+    CompletionStage<Optional<String>> fetchAndCache(@Nonnull final String key) {
+        return fetchAndCache(key, TaxCategoryQuery::of, TaxCategory::getKey, "TaxCategory");
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Set<TaxCategory>> fetchMatchingTaxCategoriesByKeys(
+        @Nonnull final Set<String> taxCategoryNames) {
+        return fetchMatchingResources(
+            taxCategoryNames,
+            () -> TaxCategoryQuery.of().withPredicates(
+                buildResourceQueryPredicate(taxCategoryNames, "key")),
```

Comment:
you can use here the JVM SDK query model DSL right away, instead of writing the helper method: `buildResourceQueryPredicate`

````java
        TaxCategoryQuery.of()
                        .plusPredicates(m -> m.key().isIn(taxCategoryKeys))
````

Reply:
in the product service this helper method was written since the JVM SDK didn't support the `key().isIn` dsl for the product query model..